### PR TITLE
JDK-8292072: NMT: repurpose Tracking overhead counter as global malloc counter

### PR DESCRIPTION
**Pull request by tstuefe**
I need a fast counter for global mallocs for [JDK-8291878](https://bugs.openjdk.org/browse/JDK-8291878). NMT does not keep a global counter, just category-specific counters. At the moment, `MallocMemorySummary::total()` iterates all these counters and adds them up, which is quite costly.

This patch makes `MallocMemorySummary::total()` fast by introducing a global malloc counter. However, we don't add a new counter, we just repurpose the tracking overhead counter. 

The tracking overhead counter gets updated on every malloc and counts number and size of all malloc headers. But malloc headers have always the same size. This patch uses the former overhead counter to count number and size of all mallocs. So now we have a global malloc counter, at no additional cost, and the overhead gets calculated from the number of total mallocs * constant header size.

---

Tests done:

- GHAs (which execute NMT jtreg + gtests tests in all modes)
- manual test (checking that tracking overhead numbers are the same as before)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292072](https://bugs.openjdk.org/browse/JDK-8292072): NMT: repurpose Tracking overhead counter as global malloc counter

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [34dfad49](https://git.openjdk.org/jdk/pull/9804/files/34dfad49a15072cc727303b3ed13c1a00f5855cb)
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9804/head:pull/9804` \
`$ git checkout pull/9804`

Update a local copy of the PR: \
`$ git checkout pull/9804` \
`$ git pull https://git.openjdk.org/jdk pull/9804/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9804`

View PR using the GUI difftool: \
`$ git pr show -t 9804`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9804.diff">https://git.openjdk.org/jdk/pull/9804.diff</a>

</details>

### CODE REVIEW

**Comment by iklam** [comment]
This looks like a good simplification. I can confirmed that the old counter was always used with a constant size, so I think your fix is correct.

```
record_new_malloc_header(sizeof(MallocHeader))
```

Have you done any testing to confirm that the old and new code reports the same overhead numbers. I see that you've moved the accounting code around so I am not sure if all the accountings are still being done at the exact same places.

Also, if possible, a new gtest would be useful.

**Review by iklam — APPROVED** [review]

**Comment by tstuefe** [comment]
Hi Ioi,

> Have you done any testing to confirm that the old and new code reports the same overhead numbers. I see that you've moved the accounting code around so I am not sure if all the accountings are still being done at the exact same places.

I checked, numbers are similar.

> 
> Also, if possible, a new gtest would be useful.

I added a gtest, but it needs quite a large leeway for heuristics to prevent false positives. Since NMT itself does malloc/free and concurrent code can malloc/free, numbers wobble quite a bit between runs. I ended up with a tolerance factor of +-30% - not sure how useful this test is apart from catching obvious errors like overflows.

**Comment by tstuefe** [comment]
May I have a second review, please?

File: src/hotspot/share/services/mallocTracker.hpp
Comment (RealCLanger) [inline src/hotspot/share/services/mallocTracker.hpp:159]: Is this correct style to implement the method in the header? Shouldn't it maybe be `inline`?

File: src/hotspot/share/services/mallocTracker.hpp
Comment (RealCLanger) [inline src/hotspot/share/services/mallocTracker.hpp:164]: Same here: Is this correct style to implement the method in the header? Shouldn't it maybe be `inline`?

**Review by RealCLanger — APPROVED** [review]
Looks good from what I can tell. I've added two remarks - not sure whether they are meaningful.

File: src/hotspot/share/services/mallocTracker.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocTracker.hpp:159]: I think it is okay for oneliners. And "inline" is implied if the definition appears in the class body.

**Comment by tstuefe** [comment]
> Looks good from what I can tell. I've added two remarks - not sure whether they are meaningful.

Thanks a lot, Christoph!

**Pull request closed without merging**
